### PR TITLE
Allow implicit string literal -> const(wchar_t)[]

### PR DESCRIPTION
**Pull request by thewilsonator**
for druntime#2349

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -3841,6 +3841,19 @@ extern (C++) final class TypeDArray : TypeArray
...
+            else if (next.ty == Tenum)
+            {
+                EnumDeclaration ed = (cast(TypeEnum)next).sym;
+                if (ed.isSpecial())
+                {
+                    /* Special enums that allow implicit conversions to them. 
+                       In particular allow string -> const(wchar_t)[]
+                     */
+                    TypeBasic tob = next.toBasetype().isTypeBasic();
+                    if (tob)
```
Comment (BBasile) [inline src/dmd/mtype.d]: you can save a LOC here :
`if (TypeBasic tob = next.toBasetype().isTypeBasic())`

**Comment by Geod24** [comment]
Could you provide some context ? I'm not sure I follow the reasoning.

**Comment by thewilsonator** [comment]
```
enum foo : wchar { whatever }

immutable(foo) [] a = "something"; // fails

enum __c_wchar_t : wchar;
immutable(__c_wchar_t)[] b = "something"; // fails but as __c_wchar_t is magic, it should pass
``` 
`__c_wchar_t` is rather a lot less useful without this.

Also where are implicit conversions tested in the test suite?

**Review by jacob-carlborg — DISMISSED** [review]
* Needs to have tests
* Needs to have a changelog entry
* Needs to have a spec update

**Comment by thewilsonator** [comment]
How is the implicit conversion from e.g. `string` to `const(char)*` for string literals done? I can't seem to see anything in `mtype.d`.

**Review by Geod24 — DISMISSED** [review]
It's not that simple. Check the following snippet:

```D
import core.stdc.stdio;
import core.stdc.string;
import core.stdc.errno;

import std.conv;
import std.stdio;

extern(C) int wprintf(const __c_wchar_t* format, ...);

enum __c_wchar_t : dchar; // Use `wchar` if you are on Windows

static immutable string foo = "Hello World   !\n";

void main ()
{
    immutable(__c_wchar_t)[] str = foo;
    if (-1 == wprintf(str.ptr))
        writeln("Error: ", strerror(errno).to!string);
}
```

This prints `Error: Illegal byte sequence" on my machine (OSX).
It's because this implicit conv is just *blindly doing an array cast* in this case.  Removing one of the character will yield an "array cast misalignment" assert error.

There is no implicit conversion between `{w,d}string` and `string`, it's just that `"foo"` is not typed as string (but `""c` is).

Side note: testing with literals is never enough.

**Comment by thewilsonator** [comment]
Ahh, right. Where is the string literal specific stuff handled? Not `mtype.d` apparently.

File: src/dmd/dcast.d [hunk 2]
```diff
@@ -650,6 +650,16 @@ MATCH implicitConvTo(Expression e, Type t)
...
                         result = m;
                         return;
+                    case Tenum:
+                        if ((cast(TypeEnum)tn).sym.isSpecial())
+                        {
+                            /* Allow string literal -> const(wchar_t)[]
+                             */
+                            if (TypeBasic tob = tn.toBasetype().isTypeBasic())
+                            result = tn.implicitConvTo(tob);
+                            return;
```
Comment (ibuclaw) [inline src/dmd/dcast.d]: Indentation here looks off. I can't tell whether the early return here should be part of the first or second condition.

**Comment by Basile-z** [comment]
Change brought by #9460  would break the test added in the PR. Have a look if you have some time to @thewilsonator .

**Pull request merged**